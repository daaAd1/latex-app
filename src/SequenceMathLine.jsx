import React from 'react';
import TextareaAutosize from 'react-autosize-textarea';

/*
**
Autor: Samuel Sepeši
Dátum: 10.5.2018
Komponent: SequenceMathLine
**
*/

/*  global localStorage: false, console: false, */

/*
Komponent, ktorý pozostáva z čiary, ktorá určuje počet potomkov, zo vstupného poľa dôkazu
a vstupného poľa podmienky pri dôkaze. Z týchto komponentov pozostáva hlavný komponent SequenceMath.
*/

/*
Komponent dostáva od rodiča - level/riadok, bunku/stĺpec, atribút, 
ktorý hovorí o tom či sa komponent zobrazí a atribút, ktorý zabraňuje upravovaniu vstupného poľa komponentu.
Komponent posiela rodičovi svoju dĺžku, keď používateľ klikne na čiaru, nový text dôkazu, 
ak bol pôvodný upravený a nový text podmienky, ak bola zmenená.
*/

class SequenceMathLine extends React.PureComponent {
  constructor(props) {
    super(props);
    this.state = {
      level: props.level,
      cell: props.cell,
      length: this.getInitialLength(),
      white: props.white,
      inputText: this.getInitialText(),
      annotation: this.getInitialAnnotation(),
      annotationText: this.getInitialAnnotationText(),
      readonlyText: props.readonlyText,
    };

    this.onClick = this.onClick.bind(this);
    this.onChange = this.onChange.bind(this);
    this.annotationChanged = this.annotationChanged.bind(this);
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.white !== this.state.white) {
      this.setState({ white: nextProps.white });
    }
    if (nextProps.readonlyAnnot !== this.state.readonlyAnnot) {
      this.setState({
        readonlyAnnot: nextProps.readonlyAnnot,
      });
    }
    if (nextProps.readonlyText !== this.state.readonlyText) {
      this.setState({
        readonlyText: nextProps.readonlyText,
      });
    }
  }

  onChange(event) {
    this.setState({
      inputText: event.target.value,
    });
    localStorage.setItem(
      `math-line-text-${this.props.level}${this.props.cell}`,
      event.target.value,
    );
    this.props.changedText(this.state.level, this.state.cell, event.target.value);
  }

  onClick() {
    if (this.state !== null && this.state.length === 3) {
      this.setState(
        {
          annotation: false,
          length: 0,
        },
        () => {
          localStorage.setItem(`math-line-length-${this.state.level}${this.state.cell}`, 0);
          localStorage.setItem(`math-line-annotation-${this.state.level}${this.state.cell}`, false);
          this.props.onClick(this.state.length);
        },
      );
    } else {
      this.setState(
        {
          annotation: true,
          length: this.state.length + 1,
        },
        () => {
          localStorage.setItem(
            `math-line-length-${this.state.level}${this.state.cell}`,
            this.state.length,
          );
          localStorage.setItem(`math-line-annotation-${this.state.level}${this.state.cell}`, true);
          this.props.onClick(this.state.length);
        },
      );
    }
  }

  getInitialText() {
    const text = localStorage.getItem(`math-line-text-${this.props.level}${this.props.cell}`) || '';
    return text;
  }

  getInitialLength() {
    const length =
      Number(localStorage.getItem(`math-line-length-${this.props.level}${this.props.cell}`)) || 0;
    return length;
  }

  getInitialAnnotation() {
    let annotation = localStorage.getItem(
      `math-line-annotation-${this.props.level}${this.props.cell}`,
    );
    if (annotation === 'true') {
      annotation = true;
    } else {
      annotation = false;
    }
    return annotation;
  }

  getInitialAnnotationText() {
    const annotationText =
      localStorage.getItem(`math-line-annotation-text-${this.props.level}${this.props.cell}`) ||
      this.props.annotationText;
    return annotationText;
  }

  annotationChanged(event) {
    this.setState({
      annotationText: event.target.value,
    });
    localStorage.setItem(
      `math-line-annotation-text-${this.state.level}${this.state.cell}`,
      event.target.value,
    );
    this.props.annotationChanged(this.state.level, this.state.cell, event.target.value);
  }

  render() {
    let className = ' ';
    let inputClassName = ' sequence-cell-text ';
    if (this.state.white) {
      className += ' sequence-cell-length-white ';
      inputClassName += ' sequence-cell-text-no-input ';
    } else if (this.state.length === 0) {
      className += ' sequence-cell-length-0 ';
    } else if (this.state.length === 1) {
      className += ' sequence-cell-length-1 ';
    } else if (this.state.length === 2) {
      className += ' sequence-cell-length-2 ';
    } else if (this.state.length === 3) {
      className += ' sequence-cell-length-3 ';
    }
    return (
      <div className="sequence-line">
        <div className="sequence-line-annotation">
          <div role="button" className={className} onClick={this.onClick} tabIndex={0} />
          {this.state.annotation && (
            <TextareaAutosize
              type="text"
              className="sequence-annotation-text"
              onChange={this.annotationChanged}
              value={this.state.annotationText}
            />
          )}
        </div>
        <TextareaAutosize
          readOnly={this.state.readonlyText}
          value={this.state.inputText}
          className={inputClassName}
          type="text"
          onChange={this.onChange}
        />
      </div>
    );
  }
}

export default SequenceMathLine;
