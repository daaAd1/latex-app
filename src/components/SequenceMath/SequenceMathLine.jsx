import React from 'react';
import TextareaAutosize from 'react-autosize-textarea';
import PropTypes from 'prop-types';

/*
**
Autor: Samuel Sepeši
Dátum: 10.5.2018
Komponent: SequenceMathLine
**
*/

/*  global console: false, */

/*
Komponent, ktorý pozostáva z čiary, ktorá určuje počet potomkov, zo vstupného poľa dôkazu
a vstupného poľa podmienky pri dôkaze. Z týchto komponentov pozostáva hlavný komponent SequenceMath.
*/

/*
Komponent dostáva od rodiča - level/riadok, bunku/stĺpec, atribút,
ktorý hovorí o tom či sa komponent zobrazí a atribút,
ktorý zabraňuje upravovaniu vstupného poľa komponentu.
Komponent posiela rodičovi svoju dĺžku, keď používateľ klikne na čiaru, nový text dôkazu,
ak bol pôvodný upravený a nový text podmienky, ak bola zmenená.
*/

class SequenceMathLine extends React.PureComponent {
  static returnInputCellClassname(white) {
    return white ? ' sequence-cell-text sequence-cell-text-no-input ' : ' sequence-cell-text ';
  }

  static returnLengthDivClassName(length, white) {
    if (white) {
      return ' sequence-cell-length-white ';
    } else if (length === 1) {
      return ' sequence-cell-length-1 ';
    } else if (length === 2) {
      return ' sequence-cell-length-2 ';
    } else if (length === 3) {
      return ' sequence-cell-length-3 ';
    }
    return ' sequence-cell-length-0 ';
  }

  constructor(props) {
    super(props);

    this.handleLengthChange = this.handleLengthChange.bind(this);
    this.handleTextChange = this.handleTextChange.bind(this);
    this.handleAnnotationChange = this.handleAnnotationChange.bind(this);
  }

  handleTextChange(event) {
    this.props.onTextChange(this.props.level, this.props.cell, event.target.value);
  }

  handleLengthChange() {
    if (this.props.length === 3) {
      this.props.onLengthChange(this.props.level, this.props.cell, 0);
    } else {
      this.props.onLengthChange(this.props.level, this.props.cell, this.props.length + 1);
    }
  }

  handleAnnotationChange(event) {
    this.props.onAnnotationChange(this.props.level, this.props.cell, event.target.value);
  }

  render() {
    const { white, inputText, length, annotation, annotationText, readonlyText } = this.props;
    const inputClassName = SequenceMathLine.returnInputCellClassname(white);
    const lengthClassName = SequenceMathLine.returnLengthDivClassName(length, white);

    return (
      <div className="sequence-line">
        <div className="sequence-line-annotation">
          <div
            role="button"
            className={lengthClassName}
            onClick={this.handleLengthChange}
            onKeyPress={this.handleLengthChange}
            tabIndex={0}
            data-testid="line-length"
          />
          {annotation && (
            <TextareaAutosize
              type="text"
              className="sequence-annotation-text"
              onChange={this.handleAnnotationChange}
              value={annotationText}
            />
          )}
        </div>
        <TextareaAutosize
          readOnly={readonlyText}
          value={inputText}
          className={inputClassName}
          type="text"
          onChange={this.handleTextChange}
        />
      </div>
    );
  }
}

SequenceMathLine.propTypes = {
  annotation: PropTypes.bool,
  inputText: PropTypes.string,
  annotationText: PropTypes.string,
  length: PropTypes.number,
  readonlyText: PropTypes.bool,
  onAnnotationChange: PropTypes.func,
  onLengthChange: PropTypes.func,
  onTextChange: PropTypes.func,

  level: PropTypes.number.isRequired,
  cell: PropTypes.number.isRequired,
  white: PropTypes.bool.isRequired,
};

SequenceMathLine.defaultProps = {
  annotation: false,
  inputText: '',
  annotationText: '',
  length: 0,
  readonlyText: false,
  onAnnotationChange: () => {},
  onLengthChange: () => {},
  onTextChange: () => {},
};

export default SequenceMathLine;
