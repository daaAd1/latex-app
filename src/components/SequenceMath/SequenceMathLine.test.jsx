import React from 'react';
import { render, renderIntoDocument, cleanup, Simulate } from 'react-testing-library';
import 'jest-dom/extend-expect';
import jest from 'jest';
import SequenceMathLine from './SequenceMathLine';

afterEach(cleanup);

test('renders without crashing', () => {
  render(<SequenceMathLine white={false} level={1} cell={1} />);
});

test('renders white cell without crashing', () => {
  const { container } = renderIntoDocument(<SequenceMathLine white level={1} cell={1} />);

  expect(container.firstChild).toMatchSnapshot();
});

test('renders with prop text and text is changed', () => {
  const { getByText } = render(
    <SequenceMathLine white={false} inputText="text" level={1} cell={1} />,
  );

  getByText('text').textContent = 'updated text';

  expect(getByText('updated text'));
});

test('renders cell, clicks length div', () => {
  const handleClick = jest.fn();
  const { getByTestId } = renderIntoDocument(
    <SequenceMathLine onClick={handleClick} white={false} inputText="text" level={1} cell={1} />,
  );

  Simulate.click(getByTestId('line-length'));

  expect(handleClick).toHaveBeenCalledTimes(1);
});
