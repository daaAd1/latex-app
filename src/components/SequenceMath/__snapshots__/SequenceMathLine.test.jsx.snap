// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders white cell without crashing 1`] = `
<div
  class="sequence-line"
>
  <div
    class="sequence-line-annotation"
  >
    <div
      role="button"
      class=" sequence-cell-length-white "
      tabindex="0"
      data-testid="line-length"
    />
  </div>
  <textarea
    class=" sequence-cell-text sequence-cell-text-no-input "
    type="text"
    rows="1"
  />
</div>
`;
